Convert LoadExtActions to a map

sparker-arm opened the pull request:
So we can store actions per address space.

RKSimon reviewed (changes requested):
(style) please don't clang-format lines unrelated to a patch

sparker-arm committed:
[NFC][Codegen] Convert LoadExtActions to a map

So we can store actions per address space. The default space, zero,
is used is one is not supplied.

sparker-arm commented:
> please don't clang-format lines unrelated to a patch

Sorry, I've removed that commit.

QuantumSegfault commented:
Behavior right now is exactly the same as before, so it works in that respect.

Maybe I'm still missing something, but I don't see how having a default parameter helps us.

Once a Custom handler is set for address space 1, the existing code will still see that the default one (0) is still "Legal", and make incorrect assumptions. So then all the (indirect) uses of getLoadExtAction (including uses isLoadExtLegal[OrCustom] of) need to be made address space aware. And then to make sure that those changes work correctly, any backends using extloads from anything other than address space 0 need fixing (not many, I imagine).

Some places are easy to get an address space for, but what about something like `BasicTTIIImpl::getCastInstrCost`?
```cpp
  InstructionCost
  getCastInstrCost(unsigned Opcode, Type *Dst, Type *Src,
                   TTI::CastContextHint CCH, TTI::TargetCostKind CostKind,
                   const Instruction *I = nullptr) const override {
```

https://github.com/llvm/llvm-project/blob/3ce16565c81df86b556cd25dc40c7a03000169dd/llvm/include/llvm/CodeGen/BasicTTIImpl.h#L1240-L1251

sparker-arm committed:
hooking up dagcombine and legalize

QuantumSegfault commented:
> Yeah, the main problem with this, current, implementation is that once you set an action for an explicit (non-zero) address space, I think this means that actions need to be set for all types :/

Actually, it goes further. If you set an action other than Legal for even addrspace 0, you have to set for every other possible address space (unless of course, the operation would be invalid) in order to maintain the current behavior.

That's why all the tests for AMDGPU were messed up. I'm looking at it, and the modified tests are all loads from non-standard address spaces. The tests had the correct behavior. The problem is that Promote and Expand are only being triggered for addrspace(0) now, not the other two or more addrspaces being tested for the behavior. It's the implementation of `AMDGPUIselLowering.cpp` that needs changing. Lines 183 - 238. Wrapping it all in a for-each loop over all the AMDGPUAS values should do it to preserve the existing behavior?

----

As for making all uses aware of address spaces, you didn't finish it ALL yet. You missed `llvm/include/llvm/CodeGen/BasicTTIIImpl.h:1284`. 

`AArch64ISelLowering` and `X86ISelLowering` also make use of `isLoadExtLegalOrCustom` and `isLoadExtLegal` respectively.

Other than those three, I think you got the rest.

----

About the crashes, looking at the logs from the test runner, it's not just the SVE test. Several other tests crash at `DAGCombiner::CombineExtLoad`

It probably comes back down to the fact that you didn't finish modifying `AMDGPUISelLowering` to account for all address spaces when running `setLoadExtAction`. I have virtually no understanding of vectors, but it's probably this that normally prevents that assertion:
```cpp
for (MVT VT : MVT::integer_fixedlen_vector_valuetypes())
      for (auto MemVT :
           {MVT::v2i8, MVT::v4i8, MVT::v2i16, MVT::v3i16, MVT::v4i16})
        setLoadExtAction({ISD::SEXTLOAD, ISD::ZEXTLOAD, ISD::EXTLOAD}, VT,
                         MemVT, Expand);
```

Should be able to ignore the crash for now.
